In [1]:
import requests
import pandas as pd

In [2]:
api_key = "dc33d51f5d112e81607b27308e5b1eb0f891774ab3937106033f7f2e452c6a69"

In [3]:
url = "https://api.football-data-api.com/league-list?key=" + api_key
response = requests.get(url)

In [4]:
data = response.json()

In [5]:
data.keys()

dict_keys(['success', 'pager', 'metadata', 'data', 'message'])

In [6]:
countries = [data.get('country') for data in data['data']]
countries

['USA',
 'Scotland',
 'Germany',
 'Europe',
 'Malaysia',
 'England',
 'Portugal',
 'Turkey',
 'England',
 'England',
 'England',
 'Spain',
 'Spain',
 'Norway',
 'Turkey',
 'Netherlands',
 'Netherlands',
 'Finland',
 'Finland',
 'France',
 'Italy',
 'France',
 'Italy',
 'Belgium',
 'Germany',
 'Germany',
 'Austria',
 'Austria',
 'Australia',
 'Brazil',
 'Brazil',
 'Sweden',
 'Sweden',
 'Argentina',
 'Argentina',
 'Argentina',
 'Poland',
 'Poland',
 'Belarus',
 'China',
 'Japan',
 'Armenia',
 'Switzerland',
 'Switzerland',
 'Azerbaijan',
 'Russia',
 'China',
 'Japan',
 'Portugal',
 'Switzerland',
 'Latvia',
 'Greece',
 'Egypt',
 'Scotland',
 'Scotland',
 'Thailand',
 'Scotland',
 'Poland',
 'Lithuania',
 'Lithuania',
 'South Africa',
 'Algeria',
 'Algeria',
 'Hong Kong',
 'Japan',
 'Armenia',
 'Albania',
 'France',
 'Mexico',
 'Mexico',
 'Norway',
 'Russia',
 'Singapore',
 'Indonesia',
 'Ghana',
 'Germany',
 'Estonia',
 'Estonia',
 'Estonia',
 'Ukraine',
 'Ukraine',
 'Ukraine',
 'Ukraine

In [7]:
countries.index("England")

5

In [8]:
seasons_info = data["data"][5]["season"]
seasons_info

[{'id': 9, 'year': 20162017},
 {'id': 10, 'year': 20152016},
 {'id': 11, 'year': 20142015},
 {'id': 12, 'year': 20132014},
 {'id': 161, 'year': 20172018},
 {'id': 246, 'year': 20122013},
 {'id': 1625, 'year': 20182019},
 {'id': 2012, 'year': 20192020},
 {'id': 3119, 'year': 20112012},
 {'id': 3121, 'year': 20102011},
 {'id': 3125, 'year': 20092010},
 {'id': 3131, 'year': 20082009},
 {'id': 3137, 'year': 20072008},
 {'id': 4759, 'year': 20202021},
 {'id': 6135, 'year': 20212022},
 {'id': 7704, 'year': 20222023},
 {'id': 9660, 'year': 20232024}]

In [9]:
season_ids = [season.get("id") for season in seasons_info]
print(season_ids)

[9, 10, 11, 12, 161, 246, 1625, 2012, 3119, 3121, 3125, 3131, 3137, 4759, 6135, 7704, 9660]


In [10]:
def get_league_matches(api_key, season_id):
    url = f"https://api.football-data-api.com/league-matches?key={api_key}&season_id={season_id}"
    response = requests.get(url)
    return response

In [11]:
def create_dataframe(api_key,season_ids):
    list_of_dfs = []
    for season_id in season_ids:
        try:
            response = get_league_matches(api_key,season_id)
            data = response.json()
            df = pd.DataFrame(data["data"])
            list_of_dfs.append(df)
        except:
            print("There was an error.")
            exit()
     # Concatenate the dataframes in the list
    concatenated_df = pd.concat(list_of_dfs, ignore_index=True)
    return concatenated_df

In [12]:
matches_df = create_dataframe(api_key,season_ids)
matches_df.head()

id  homeID  awayID     season    status  roundID  game_week  \
0  2155     150     108  2016/2017  complete       19          1   
1  2156     145     154  2016/2017  complete       19          1   
2  2157     143     142  2016/2017  complete       19          1   
3  2158     144      92  2016/2017  complete       19          1   
4  2159     147     141  2016/2017  complete       19          1   

   revised_game_week   homeGoals awayGoals  ...  matches_completed_minimum  \
0                 -1  [45+1, 57]      [47]  ...                         38   
1                 -1          []      [82]  ...                         38   
2                 -1          []      [74]  ...                         38   
3                 -1         [5]      [59]  ...                         38   
4                 -1        [11]      [67]  ...                         38   

   over05  over15  over25  over35  over45  over55   btts  homeGoals_timings  \
0    True    True    True   False   False   False   True         [45+1, 57]   
1    True   False   False   False   False   False  False                 []   
2    True   False   False   False   False   False  False                 []   
3    True    True   False   False   False   False   True                [5]   
4    True    True   False   False   False   False   True               [11]   

   awayGoals_timings  
0               [47]  
1               [82]  
2               [74]  
3               [59]  
4               [67]  

[5 rows x 215 columns]

In [13]:
matches_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6460 entries, 0 to 6459
Columns: 215 entries, id to awayGoals_timings
dtypes: bool(7), float64(80), int64(112), object(16)
memory usage: 10.3+ MB


In [15]:
data['pager']

{'current_page': 1,
 'max_page': 1,
 'results_per_page': 2000,
 'total_results': 1498}

In [16]:
data['metadata']

{'request_limit': '1800',
 'request_remaining': '1797',
 'request_reset_message': 'Request limit is refreshed every hour.'}

In [17]:
# Check for null values in each column for null column
null_columns = matches_df.columns[matches_df.isnull().all()].tolist()
null_columns

[]

In [18]:
# Check for null values in each column
null_columns = matches_df.columns[matches_df.isnull().any()].tolist()
null_columns

['refereeID', 'coach_a_ID', 'coach_b_ID']

In [19]:
(matches_df.refereeID.isnull().sum()/len(matches_df.refereeID)) * 100

43.01857585139319

In [20]:
(matches_df.coach_a_ID.isnull().sum()/len(matches_df.refereeID)) * 100

0.23219814241486067

In [21]:
(matches_df.coach_b_ID.isnull().sum()/len(matches_df.refereeID)) * 100

0.23219814241486067

In [22]:
print(matches_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6460 entries, 0 to 6459
Columns: 215 entries, id to awayGoals_timings
dtypes: bool(7), float64(80), int64(112), object(16)
memory usage: 10.3+ MB
None


In [23]:
columns_to_drop = ["homeGoals","awayGoals","home_url","home_image",
                   "home_name","away_url","away_image","match_url",
                   "homeGoals_timings","awayGoals_timings"

]
matches_df.drop(columns = columns_to_drop ,axis = 1,inplace = True)

We can drop Identifiers and URLs such as;

id, homeID, awayID, refereeID, coach_a_ID, coach_b_ID, match_url, home_url, home_image, away_url, away_image

In [24]:
# Check for duplicates across all columns
duplicates = matches_df[matches_df.duplicated()]
duplicates

Empty DataFrame
Columns: [id, homeID, awayID, season, status, roundID, game_week, revised_game_week, homeGoalCount, awayGoalCount, totalGoalCount, team_a_corners, team_b_corners, totalCornerCount, team_a_offsides, team_b_offsides, team_a_yellow_cards, team_b_yellow_cards, team_a_red_cards, team_b_red_cards, team_a_shotsOnTarget, team_b_shotsOnTarget, team_a_shotsOffTarget, team_b_shotsOffTarget, team_a_shots, team_b_shots, team_a_fouls, team_b_fouls, team_a_possession, team_b_possession, refereeID, coach_a_ID, coach_b_ID, stadium_name, stadium_location, team_a_cards_num, team_b_cards_num, odds_ft_1, odds_ft_x, odds_ft_2, odds_ft_over05, odds_ft_over15, odds_ft_over25, odds_ft_over35, odds_ft_over45, odds_ft_under05, odds_ft_under15, odds_ft_under25, odds_ft_under35, odds_ft_under45, odds_btts_yes, odds_btts_no, odds_team_a_cs_yes, odds_team_a_cs_no, odds_team_b_cs_yes, odds_team_b_cs_no, odds_doublechance_1x, odds_doublechance_12, odds_doublechance_x2, odds_1st_half_result_1, odds_1st_half_result_x, odds_1st_half_result_2, odds_2nd_half_result_1, odds_2nd_half_result_x, odds_2nd_half_result_2, odds_dnb_1, odds_dnb_2, odds_corners_over_75, odds_corners_over_85, odds_corners_over_95, odds_corners_over_105, odds_corners_over_115, odds_corners_under_75, odds_corners_under_85, odds_corners_under_95, odds_corners_under_105, odds_corners_under_115, odds_corners_1, odds_corners_x, odds_corners_2, odds_team_to_score_first_1, odds_team_to_score_first_x, odds_team_to_score_first_2, odds_win_to_nil_1, odds_win_to_nil_2, odds_1st_half_over05, odds_1st_half_over15, odds_1st_half_over25, odds_1st_half_over35, odds_1st_half_under05, odds_1st_half_under15, odds_1st_half_under25, odds_1st_half_under35, odds_2nd_half_over05, odds_2nd_half_over15, odds_2nd_half_over25, odds_2nd_half_over35, odds_2nd_half_under05, odds_2nd_half_under15, odds_2nd_half_under25, ...]
Index: []

[0 rows x 205 columns]

In [34]:
matches_df["stadium_location"].isna().sum()

0

In [25]:
matches_df["stadium_location"][0]

''

In [26]:
len(matches_df[matches_df["stadium_location"] == '' ])

3757

Do we need the stadium location? It has 3757 rows will null values encoded as ''

attendance column has null values encoded as -1

In [27]:
len(matches_df[matches_df["attendance"] == -1])

1822

It seems like this columns have null values encoded as -1

corner_timings_recorded
card_timings_recorded	
team_a_fh_corners	
team_b_fh_corners	
team_a_2h_corners	
team_b_2h_corners	
corner_fh_count	
corner_2h_count

Do we need this and for what purpose?

revised_game_week

date_unix

goalTimingDisabled

corner_timings_recorded

card_timings_recorded

no_home_away


In [33]:
print(matches_df.dtypes)

id         int64
homeID     int64
awayID     int64
season    object
status    object
           ...  
over25      bool
over35      bool
over45      bool
over55      bool
btts        bool
Length: 205, dtype: object


Work on boolean values?